File: R/run_pySCENIC.R
#' Uses reticulate package to run pySCENIC from R
#'
#' Run pySCENIC pipeline
#' @param dir directory path to output pySCENIC results
#' @param anndata_path path to anndata object
#' @param loom_path path to write initial loom file
#' @param tfs_path path to list of TFs
#' @param rank_db_path path to ranking database
#' @param motif_path path to motif list
#' @param output_loom_path path to write output loom file
#' @keywords pySCENIC
#' @export
#' @import reticulate
#' @examples
#' run_pyscenic(dir, anndata_path, loom_path, tfs_path, rank_db_path, motif_path, output_loom_path)


run_pyscenic <- function(dir, anndata_path, loom_path, tfs_path, rank_db_path, motif_path, output_loom_path) {
  run_pyscenic_path <- system.file("python","run_pySCENIC.py", package="scRegNetwrapper")
  source_python(run_pyscenic_path)
  pyscenic_wrapper(dir, anndata_path, loom_path, tfs_path, rank_db_path, motif_path, output_loom_path)
}


#' Uses reticulate package to set up loom file for pySCENIC pipeline
#'
#' Run GRN from pySCENIC
#' @param anndata_path path to anndata object
#' @param loom_path path to write initial loom file
#' @keywords pySCENIC loom set up
#' @export
#' @import reticulate
#' @examples
#' run_loom_setup(anndata_path, loom_path)


run_loom_setup <- function(anndata_path, loom_path){
  setup_loom_wrapper(anndata_path, loom_path)
}

#' Uses reticulate package to run GRN from R
#'
#' Run GRN from pySCENIC pipeline
#' @param dir directory path to output pySCENIC results
#' @param loom_path path to write initial loom file
#' @param tfs_path path to list of TFs
#' @param adjacencies_fname filename of adjacencies output
#' @keywords GRN
#' @export
#' @import reticulate
#' @examples
#' run_grn(dir, loom_path, tfs_path,adjacencies_fname)

run_grn <- function(dir, loom_path, tfs_path, adjacencies_fname) {
  grn_wrapper(dir, loom_path, tfs_path, adjacencies_fname)
}


#' Uses reticulate package to run CisTarget (regulon prediction) from R
#'
#' Run pySCENIC pipeline
#' @param dir directory path to output pySCENIC results
#' @param adjacencies_fname filename of adjacencies output
#' @param loom_path path to write initial loom file
#' @param rank_db_path path to ranking database
#' @param motif_path path to motif list
#' @param regulons_fname filename of regulons output
#' @keywords CisTarget
#' @export
#' @import reticulate
#' @examples
#' run_cistarget(dir, adjacencies_fname, loom_path, rank_db_path, motif_path, regulons_fname)


run_cistarget <- function(dir, adjacencies_fname, loom_path, rank_db_path, motif_path, regulons_fname) {
  cistarget_wrapper(dir, adjacencies_fname, loom_path, rank_db_path, motif_path, regulons_fname)
}

#' Uses reticulate package to run AUCell from R
#'
#' Run AUCell from pySCENIC pipeline
#' @param dir directory path to output pySCENIC results
#' @param regulons_fname filename of regulons output
#' @param loom_path path to write initial loom file
#' @param output_loom_path path to write output loom file
#' @keywords pySCENIC
#' @export
#' @import reticulate
#' @examples
#' run_aucell(dir, regulons_fname, loom_path, output_loom_path)

run_aucell <- function(dir, regulons_fname, loom_path, output_loom_path) {
  aucell_wrapper(dir, regulons_fname, loom_path, output_loom_path)
}

File: inst/python/run_pySCENIC.py
import os
import subprocess
import numpy as np
import pandas as pd
import scanpy as sc
import loompy as lp
import glob
import json
import base64
import zlib
import matplotlib.pyplot as plt
import matplotlib as mpl
from pyscenic.utils import modules_from_adjacencies, load_motifs
from pyscenic.rnkdb import FeatherRankingDatabase as RankingDatabase
from dask.diagnostics import ProgressBar
from arboreto.utils import load_tf_names
from arboreto.algo import grnboost2
from pyscenic.cli.utils import load_signatures
from pyscenic.prune import prune2df, df2regulons
from pyscenic.aucell import aucell
 
def pyscenic_wrapper(dir, anndata_path, loom_path, tfs_path, rank_db_path, motif_path, output_loom_path):
    anndata = sc.read(anndata_path,         
    var_names='gene_symbols',
    cache=False)
    anndata.var_names_make_unique()
    anndata = anndata[anndata.obs['donor'] == 'nst9']
    # create basic row and column attributes for the loom file:
    row_attrs = {
        "Gene": np.array(anndata.var_names) ,
    }
    col_attrs = {
        "CellID": np.array(anndata.obs_names) ,
        "nGene": np.array( np.sum(anndata.X.transpose()>0 , axis=0)).flatten() ,
        "nUMI": np.array( np.sum(anndata.X.transpose() , axis=0)).flatten() ,
    }
    lp.create( loom_path, anndata.X.transpose(), row_attrs, col_attrs)
    
    os.chdir(dir)
    ##### SCENIC #### 
    # STEP 1: GRN BOOST
    cmd1 = "/usr/local/bin/pyscenic grn " + loom_path + " " + tfs_path + " -o anndata_adj.csv --num_workers 20"
    os.system(cmd1)

    # STEP 2/3: Regulon prediction aka cisTarget from CL
    # ranking databases
    db_names = ' '.join( glob.glob(rank_db_path) )
    
    # running it (prunes regulons for which < 80% genes can be matched to the ranking database)
    cmd2 = "/usr/local/bin/pyscenic grn anndata_adj.csv " + db_names + " --annotations_fname " + motif_path + " --expression_mtx_fname " + loom_path + " --output anndata_reg.csv --mask_dropouts --num_workers 20"
    os.system(cmd2)

    # STEP 4: AUCell
    cmd3 = "/usr/local/bin/pyscenic aucell " + loom_path + " anndata_reg.csv --output " + output_loom_path + " --num_workers 20"
    os.system(cmd3)
  
  
def setup_loom_wrapper(anndata_path, loom_path):
    anndata = sc.read(anndata_path,         
    var_names='gene_symbols',
    cache=False)
    anndata.var_names_make_unique()
    anndata = anndata[anndata.obs['donor'] == 'nst9']
    # create basic row and column attributes for the loom file:
    row_attrs = {
        "Gene": np.array(anndata.var_names) ,
    }
    col_attrs = {
        "CellID": np.array(anndata.obs_names) ,
        "nGene": np.array( np.sum(anndata.X.transpose()>0 , axis=0)).flatten() ,
        "nUMI": np.array( np.sum(anndata.X.transpose() , axis=0)).flatten() ,
    }
    lp.create( loom_path, anndata.X.transpose(), row_attrs, col_attrs)
    
def grn_wrapper(dir, loom_path, tfs_path):
    os.chdir(dir)
    ##### SCENIC #### 
    # STEP 1: GRN BOOST
    cmd1 = "/usr/local/bin/pyscenic grn " + loom_path + " " + tfs_path + " -o anndata_adj.csv --num_workers 20"
    os.system(cmd1)
    
def cistarget_wrapper(dir, loom_path, rank_db_path, motif_path):
    os.chdir(dir)
   # STEP 2/3: Regulon prediction aka cisTarget from CL
    # ranking databases
    db_names = ' '.join( glob.glob(rank_db_path) )
    
    # running it (prunes regulons for which < 80% genes can be matched to the ranking database)
    cmd2 = "/usr/local/bin/pyscenic grn anndata_adj.csv " + db_names + " --annotations_fname " + motif_path + " --expression_mtx_fname " + loom_path + " --output anndata_reg.csv --mask_dropouts --num_workers 20"
    os.system(cmd2)
    
def aucell_wrapper(dir, loom_path, output_loom_path):
    # STEP 4: AUCell
    os.chdir(dir)
    cmd3 = "/usr/local/bin/pyscenic aucell " + loom_path + " anndata_reg.csv --output " + output_loom_path + " --num_workers 20"
    os.system(cmd3)
    
    
